hub/core/chunk.py

```diff
@@ -0,0 +1,141 @@
+        encoded_byte_positions: np.ndarray = None,
+        data: memoryview = None,
+    ):
+        """Blob storage of bytes. Tensors are stored as chunks. Tensor data is split into chunks of roughly the same size.
+        `ChunkEngine` handles the creation of `Chunk`s and the delegation of samples into them.
+
+        Data layout:
+            Every chunk has data and a header.
+
+            Data:
+                All samples this chunk contains are added into `_data` in bytes form directly adjacent to one another, without
+                delimeters.
+
+            Header:
+                All samples this chunk contains need 2 components: shape and byte position.
+                `ShapeEncoder` handles encoding the `start_byte` and `end_byte` for each sample.
+                `BytePositionsEncoder` handles encoding the `shape` for each sample.
+
+            To see how this chunk is composed into bytes, check out `tobytes`.
+            To see how this chunk is constructed from bytes, checkout `frombuffer`.
+
+        Args:
+            encoded_shapes (np.ndarray): Used to construct `ShapeEncoder` if this chunk already exists. Defaults to None.
+            encoded_byte_positions (np.ndarray): Used to construct `BytePositionsEncoder` if this chunk already exists. Defaults to None.
+            data (memoryview, optional): If this chunk already exists, data should be set. Defaults to None.
+        """
+
+        # data
+        self._data: Union[bytearray, memoryview] = data or bytearray()
+
+        # header
+        self.shapes_encoder = ShapeEncoder(encoded_shapes)
+        self.byte_positions_encoder = BytePositionsEncoder(encoded_byte_positions)
+
+    @property
+    def memoryview_data(self):
+        return memoryview(self._data)
+
+    @property
+    def num_data_bytes(self):
+        return len(self._data)
+
+    def is_under_min_space(self, min_data_bytes_target: int):
+        return self.num_data_bytes < min_data_bytes_target
+
+    def has_space_for(self, num_bytes: int, max_data_bytes: int):
+        return self.num_data_bytes + num_bytes <= max_data_bytes
+
+    def append_sample(
+        self, incoming_buffer: memoryview, max_data_bytes: int
+    ) -> Tuple["Chunk"]:
+        """Store `incoming_buffer` in this chunk.
+
+        Args:
+            incoming_buffer (memoryview): Buffer that represents a single sample.
+            max_data_bytes (int): Used to determine if this chunk has space for `incoming_buffer`.
+
+        Raises:
+            FullChunkError: If `incoming_buffer` is too large.
+        """
+
+        incoming_num_bytes = len(incoming_buffer)
+
+        if not self.has_space_for(incoming_num_bytes, max_data_bytes):
+            raise FullChunkError(
+                f"Chunk does not have space for the incoming bytes (incoming={incoming_num_bytes}, max={max_data_bytes})."
+            )
+
+        # note: incoming_num_bytes can be 0 (empty sample)
+        self._data += incoming_buffer
```

Comment:
Is this a copy? In the future we may want to do some more sophisticated reference tracking to minimize these operations.